modules/gpgsqlbackend/gpgsqlbackend.cc

```diff
@@ -98,6 +98,7 @@ class gPgSQLFactory : public BackendFactory
     declare(suffix, "id-query", "Basic with ID query", record_query+" disabled=false and type=$1 and name=$2 and domain_id=$3");
     declare(suffix, "any-query", "Any query", record_query+" disabled=false and name=$1");
     declare(suffix, "any-id-query", "Any with ID query", record_query+" disabled=false and name=$1 and domain_id=$2");
+    declare(suffix, "closest-soa-query", "Closest SOA query", "SELECT content,ttl,prio,type,domain_id,disabled::int,name,auth::int FROM records WHERE disabled=false and type='SOA' and (name=$1 or $2 LIKE '%.'||name) ORDER BY LENGTH(name) DESC");
```

Comment:
Ah yes, I've written something similar yesterday for PostgreSQL, where I use something like this:

```
select u.name, count(records.*) from unnest($1::text[]) as u(name) left join records on u.name = records.name group by 1 order by length(u.name) desc
```

But I didn't spend time on making this work properly yet with parameter binding and I'm not sure how to do this for MySQL or SQLite. I guess we can teach the sxxsql interfaces more smarts and pass a dynamic list of parameters ($1=a.b.c.example.com, $2=b.c.example.com, $3=c.example.com, ...) but it's not going to be "nice" :)

Reply:
I have a working code for spgsql here[1], if you are interested :-) But I don't think it can work in generic SQL.

[1]: https://github.com/PowerDNS/pdns/compare/master...rgacogne:auth-getbestauth3#diff-006c681559a925e671f4640858881b3a

Reply:
Looking at the rest of your code, I like the spgsql cleanup and the getAuth/getBestAuth "name split" a lot  ❤️ 

For people following along quietly, my meh-ish array binding implementation lives around here: https://github.com/zeha/pdns/blob/ab19caf7703e13d17637d6688a13877e23bbfbc2/modules/gpgsqlbackend/gpgsqlbackend.cc#L219